### PR TITLE
Add OWLv2, bis

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds OWLv2 in a way that is more compliant to the Transformers-philosophy, meaning one paper = one model = one file. Rather than modifying the existing OWL-ViT (v1), this PR adds a new standalone Owlv2ForObjectDetection model. It copies 99% of the v1 model, only modifying the object detection head.

Follow-up of #26379 

To do:

- [x] fix image processor
- [x] add doc tests
- [x] verify objectness logits shape

### CODE REVIEW

**Comment by sgvaze** [comment]
Hi, any idea when this PR will land? Eagerly waiting on playing with the model in HF!

File: docs/source/en/model_doc/owlv2.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+Tips:
+
+- Usage of OWLv2 is identical to [OWL-ViT](owlvit) with a new, updated image processor.
+
+This model was contributed by [nielsr](https://huggingface.co/nielsr).
+The original code can be found [here](https://github.com/google-research/scenic/tree/main/scenic/projects/owl_vit).
+
+
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/owlv2.md:35]: Let's add a small snipper here!

File: docs/source/en/model_doc/owlv2.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+rendered properly in your Markdown viewer.
+
+-->
+
+# OWLv2
+
+## Overview
+
+OWLv2 was proposed in [Scaling Open-Vocabulary Object Detection](https://arxiv.org/abs/2306.09683)
+by Matthias Minderer, Alexey Gritsenko, Neil Houlsby. OWLv2 scales up [OWL-ViT](owlvit) using self-training, which uses an existing detector to generate pseudo-box annotations on image-text pairs. This results in large gains over the previous state-of-the-art for zero-shot object detection.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/owlv2.md:22]:
```suggestion
OWLv2 was proposed in [Scaling Open-Vocabulary Object Detection](https://arxiv.org/abs/2306.09683) by Matthias Minderer, Alexey Gritsenko, Neil Houlsby. OWLv2 scales up [OWL-ViT](owlvit) using self-training, which uses an existing detector to generate pseudo-box annotations on image-text pairs. This results in large gains over the previous state-of-the-art for zero-shot object detection.
```

File: src/transformers/models/owlv2/configuration_owlv2.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+            [`Owlv2Config`]: An instance of a configuration object
+        """
+        config_dict = {}
+        config_dict["text_config"] = text_config
+        config_dict["vision_config"] = vision_config
+
+        return cls.from_dict(config_dict, **kwargs)
+
+
+class Owlv2OnnxConfig(OnnxConfig):
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/configuration_owlv2.py:342]: I don't we use this anymore in the transformers library (lives in optimum)

File: src/transformers/models/owlv2/convert_owlv2_to_hf.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    model = Owlv2ForObjectDetection(config)
+    missing_keys, unexpected_keys = model.load_state_dict(state_dict, strict=False)
+    print("Missing keys:", missing_keys)
+    print("Unexpected keys:", unexpected_keys)
+    assert missing_keys == ["owlv2.visual_projection.weight"]
+    assert unexpected_keys == [] if "v2" in model_name else ["class_head/padding", "class_head/padding_bias"]
+    model.eval()
+
+    # for name, param in model.named_parameters():
+    #     print(name, param.shape)
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/convert_owlv2_to_hf.py:241]:
```suggestion
```

File: src/transformers/models/owlv2/convert_owlv2_to_hf.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    for src, dest in rename_keys:
+        rename_and_reshape_key(state_dict, src, dest, config)
+
+    # load HuggingFace model
+    model = Owlv2ForObjectDetection(config)
+    missing_keys, unexpected_keys = model.load_state_dict(state_dict, strict=False)
+    print("Missing keys:", missing_keys)
+    print("Unexpected keys:", unexpected_keys)
+    assert missing_keys == ["owlv2.visual_projection.weight"]
+    assert unexpected_keys == [] if "v2" in model_name else ["class_head/padding", "class_head/padding_bias"]
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/convert_owlv2_to_hf.py:237]: I thin this might be confusing we should add in the proint that this is expected maybe?

File: src/transformers/models/owlv2/convert_owlv2_to_hf.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    Copy/paste/tweak model's weights to our OWL-ViT structure.
+    """
+    config = get_owlv2_config(model_name)
+
+    # Load original state dict based on model name
+    model_name_to_checkpoint_path = {
+        "owlvit-base-patch16": "/Users/nielsrogge/Documents/OWL-ViT/clip_vit_b16_6171dab",
+        "owlv2-base-patch16": "/Users/nielsrogge/Documents/OWL-ViT/owl2-b16-960-st-ngrams_c7e1b9a",
+        "owlv2-base-patch16-ensemble": "/Users/nielsrogge/Documents/OWL-ViT/owl2-b16-960-st-ngrams-curated-ft-lvisbase-ens-cold-weight-05_209b65b",
+    }
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/convert_owlv2_to_hf.py:216]: MMmm we should try to be agnostic to your path and use an input path to the checkoints

File: src/transformers/models/owlv2/convert_owlv2_to_hf.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    with torch.no_grad():
+        outputs = model(input_ids=input_ids, pixel_values=pixel_values)
+        logits = outputs.logits
+        pred_boxes = outputs.pred_boxes
+
+    # # note: the logits below were obtained without center cropping
+    # if checkpoint_url == "https://dl.fbaipublicfiles.com/convnext/owlvit/im1k/owlvit_atto_1k_224_ema.pt":
+    #     expected_logits = torch.tensor([-0.3930, 0.1747, -0.5246, 0.4177, 0.4295])
+    # else:
+    #     raise ValueError(f"Unknown URL: {checkpoint_url}")
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/convert_owlv2_to_hf.py:270]: To remove?

File: src/transformers/models/owlv2/convert_owlv2_to_hf.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+            [[-8.6353, -9.5409, -6.6154], [-7.9442, -9.6151, -6.7117], [-12.4593, -15.3332, -12.1048]]
+        )
+        expected_boxes = torch.tensor([[0.0126, 0.0090, 0.0238], [0.0387, 0.0227, 0.0754], [0.0582, 0.1058, 0.1139]])
+
+    print("Logits:", logits[0, :3, :3])
+    print("Pred boxes:", pred_boxes[0, :3, :3])
+
+    assert torch.allclose(logits[0, :3, :3], expected_logits, atol=1e-3)
+    assert torch.allclose(pred_boxes[0, :3, :3], expected_boxes, atol=1e-3)
+    print("Looks ok!")
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/convert_owlv2_to_hf.py:293]: Thanks for updating! 
I still think we should allow to disable this test with a kwarg, but if indeed the framework used fo owl is not used a lot / won't produce other checkpoints then no worries!

File: src/transformers/models/owlv2/image_processing_owlv2.py [hunk 9]
```diff
@@ -0,0 +1,608 @@
...
+    elif output_ndim < image.ndim:
+        raise ValueError("output_shape length cannot be smaller than the " "image number of dimensions")
+
+    return image, output_shape
+
+
+def _clip_warp_output(input_image, output_image, mode, cval):
+    """Clip output image to range of values of input image.
+
+    Note that this function modifies the values of *output_image* in-place and it is only modified if `clip=True`.
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/image_processing_owlv2.py:149]: I don't really know where the `clip=True` argument comes from?

File: src/transformers/models/owlv2/image_processing_owlv2.py [hunk 10]
```diff
@@ -0,0 +1,608 @@
...
+            images (`ImageInput`):
+                Image to preprocess. Expects a single or batch of images with pixel values ranging from 0 to 255. If
+                passing in images with pixel values between 0 and 1, set `do_rescale=False`.
+            do_rescale (`bool`, *optional*, defaults to `self.do_rescale`):
+                Whether to rescale the image values between [0 - 1].
+            rescale_factor (`float`, *optional*, defaults to `self.rescale_factor`):
+                Rescale factor to rescale the image by if `do_rescale` is set to `True`.
+            do_pad (`bool`, *optional*, defaults to `self.do_pad`):
+                Whether to pad the image to a square with gray pixels on the bottom and the right.
+            do_resize (`bool`, *optional*, defaults to `self.do_resize`):
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/image_processing_owlv2.py:374]: order is a bit different but its a nit

File: src/transformers/models/owlv2/image_processing_owlv2.py [hunk 11]
```diff
@@ -0,0 +1,608 @@
...
+                        "Anti-aliasing standard deviation greater than zero but " "not down-sampling along all axes"
+                    )
+            filtered = ndi.gaussian_filter(image, anti_aliasing_sigma, cval=cval, mode=ndi_mode)
+        else:
+            filtered = image
+
+        zoom_factors = [1 / f for f in factors]
+        out = ndi.zoom(filtered, zoom_factors, order=order, mode=ndi_mode, cval=cval, grid_mode=True)
+
+        _clip_warp_output(image, out, mode, cval)
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/image_processing_owlv2.py:339]: bit strange that everything else is not inplace but this is no?

File: src/transformers/models/owlv2/image_processing_owlv2.py [hunk 12]
```diff
@@ -0,0 +1,608 @@
...
+        """
+        Resize an image as per the original implementation.
+
+        Args:
+            image (`np.ndarray`):
+                Image to resize.
+            size (`Dict[str, int]`):
+                Dictionary containing the height and width to resize the image to.
+            data_format (`str` or `ChannelDimension`, *optional*):
+                The channel dimension format of the image. If not provided, it will be the same as the input image.
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/image_processing_owlv2.py:303]: missing 2 args

File: src/transformers/models/owlv2/modeling_owlv2.py [hunk 13]
```diff
@@ -0,0 +1,1778 @@
...
+        """Predicts the probability that each image feature token is an object.
+        Args:
+            image_features (`torch.FloatTensor` of shape `(batch_size, num_patches, hidden_dim)`)):
+                Features extracted from the image.
+        Returns:
+            Objectness scores.
+        """
+        # TODO support training mode
+        # if self.objectness_head_configs.stop_gradient:
+        #     image_features = jax.lax.stop_gradient(image_features)
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/modeling_owlv2.py:1389]: Seems like this should be remove!

File: src/transformers/models/owlv2/processing_owlv2.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team.
```
Comment (ArthurZucker) [inline src/transformers/models/owlv2/processing_owlv2.py:2]:
```suggestion
# Copyright 2023 The HuggingFace Inc. team.
```

File: src/transformers/models/owlvit/modeling_owlvit.py
Comment (ArthurZucker) [inline src/transformers/models/owlvit/modeling_owlvit.py:1]: Good for me!

File: tests/models/owlv2/test_image_processor_owlv2.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+        image_processor = self.image_processing_class.from_dict(
+            self.image_processor_dict, size={"height": 42, "width": 42}
+        )
+        self.assertEqual(image_processor.size, {"height": 42, "width": 42})
+
+    @slow
+    def test_image_processor_integration_test(self):
+        processor = Owlv2ImageProcessor()
+
+        filepath = hf_hub_download(repo_id="adirik/OWL-ViT", repo_type="space", filename="assets/astronaut.png")
```
Comment (ArthurZucker) [inline tests/models/owlv2/test_image_processor_owlv2.py:119]: Should probably use hf internal testing rather than a private repo

File: tests/models/owlv2/test_modeling_owlv2.py
Comment (ArthurZucker) [inline tests/models/owlv2/test_modeling_owlv2.py:1]: Just missing some copied from here!

**Review by ArthurZucker — APPROVED** [review]
Thanks for adding the copied from mostly everywhere! Just missing in the model testing 😉 
LGTM

File: docs/source/en/model_doc/owlv2.md (hunk 1 above)
Comment (NielsRogge) [reply docs/source/en/model_doc/owlv2.md:35]: Typically we add a code snippet (that is also doc tested) below the forward of every model. But I can add an additional one here

**Review by ArthurZucker — APPROVED** [review]
Thanks for adding this new model 🚀 waiting for the cI to merge

**Comment by NielsRogge** [comment]
Thanks for spotting @assafbot would you be able to open a PR regarding this?
